src/mscorlib/src/System/Lazy.cs

```diff
@@ -322,147 +505,13 @@ public T Value
         {
             get
             {
-                Boxed boxed = null;
-                if (m_boxed != null )
-                {
-                    // Do a quick check up front for the fast path.
-                    boxed = m_boxed as Boxed;
-                    if (boxed != null)
-                    {
-                        return boxed.m_value;
-                    }
-
-                    LazyInternalExceptionHolder exc = m_boxed as LazyInternalExceptionHolder;
-                    Debug.Assert(exc != null);
-                    exc.m_edi.Throw();
-                }
+                var implementation = m_implementation;
+                if (implementation == null)
+                    return m_value;
 
-                // Fall through to the slow path.
-                return LazyInitValue();
+                return implementation.GetValue(this);
```

Comment:
Have you tried moving `implementation.GetValue(this)` into separate method, like the LazyInitValue that was there before? It may make the already-initialized path faster.